rclcpp/include/rclcpp/serialization.hpp

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__SERIALIZATION_HPP_
+#define RCLCPP__SERIALIZATION_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rosidl_typesupport_cpp/message_type_support.hpp"
+
+#include "rcl/types.h"
+#include "rosidl_runtime_c/message_type_support_struct.h"
+
+namespace rclcpp
+{
+
+class SerializedMessage;
+
+/// Interface to (de)serialize a message
+class SerializationBase
+{
+protected:
+  explicit SerializationBase(const rosidl_message_type_support_t * type_support);
+
+  virtual ~SerializationBase() = default;
+
+  /// Serialize a ROS2 message to a serialized stream
+  /**
+   * \param[in] message The ROS2 message which is read and serialized by rmw.
+   * \param[out] serialized_message The serialized message.
+   */
+  void serialize_message(
+    const void * ros_message, SerializedMessage * serialized_message) const;
+
+  /// Deserialize a serialized stream to a ROS message
+  /**
+   * \param[in] serialized_message The serialized message to be converted to ROS2 by rmw.
+   * \param[out] message The deserialized ROS2 message.
+   */
+  void deserialize_message(
+    const SerializedMessage * serialized_message, void * ros_message) const;
+
+  const rosidl_message_type_support_t * type_support_;
+};
+
+/// Default implementation to (de)serialize a message by using rmw_(de)serialize
+template<typename MessageT>
+class Serialization : public SerializationBase
```

Comment:
The base class is template independent. That's going to be necessary later in the context of rosbags, where the `rosidl_typesupport_cpp` handle is being loaded on runtime.

Having a template argument on the derived classes makes sense to me as it will make that class to get the handle opaque. Mainly syntactic sugar I guess.

Reply:
> This class will also inherit the serialize_message and deserialize_message signatures from the base class.

That's a fair point. @DensoADAS was that intended to have both functions (one which takes pointers, one which takes references)? In this case, I actually think it's fine to take `const void *` to the ros messages.

Reply:
> That's a fair point. @DensoADAS was that intended to have both functions (one which takes pointers, one which takes references)?
It was the result by taking the original code. Not intended.

Reply:
> The base class is template independent. That's going to be necessary later in the context of rosbags, where the rosidl_typesupport_cpp handle is being loaded on runtime.

Yes, what I mean is that the base class isn't providing a base interface for the others, so it just seem to be a different class.

Reply:
Agreed.

I think this could be one class with two constructors, one which takes the rosidl_typesupport handle and one templated constructor which deducts the handle based on the template.